language/stdlib/modules/option.move

```diff
@@ -0,0 +1,119 @@
+module Option {
+    use 0x0::Vector;
+
+    // Abstraction of a value that may or may not be present. Implemented with a vector of size
+    // zero or one because Move bytecode does not have ADTs.
+    struct T<Element> { vec: vector<Element> }
+
+    // Return an empty `Option::T`
+    public fun none<Element>(): T<Element> {
+        T { vec: Vector::empty() }
+    }
+
+    // Return an `Option::T` containing `e`
+    public fun some<Element>(e: Element): T<Element> {
+        T { vec: Vector::singleton(e) }
+    }
+
+    // Return true if `t` does not hold a value
+    public fun is_none<Element>(t: &T<Element>): bool {
+        Vector::is_empty(&t.vec)
+    }
+
+    // Return true if `t` holds a value
+    public fun is_some<Element>(t: &T<Element>): bool {
+        !Vector::is_empty(&t.vec)
+    }
+
+    // Return true if the value in `t` is equal to `e_ref`
+    // Always returns `false` if `t` does not hold a value
+    public fun contains<Element>(t: &T<Element>, e_ref: &Element): bool {
+        Vector::borrow(&t.vec, 0) == e_ref
+    }
+
+    // Return an immutable reference to the value inside `t`
+    // Aborts if `t` does not hold a value
+    public fun borrow<Element>(t: &T<Element>): &Element {
+        Vector::borrow(&t.vec, 0)
+    }
+
+    // TODO: not clear this API is a good idea
+    // Return a reference to the value inside `t` if it holds one
+    // Return `default_ref` if `t` does not hold a value
+    public fun borrow_default<Element>(t: &T<Element>, default_ref: &Element): &Element {
```

Comment:
I think having this API makes sense in the presence of resources. Additionally, this is a much more general API that has as a special case `get_default` -- the user can decide if they need to/can copy into an owned value after `borrow_default`ing based upon their local kind restrictions on the type held within the `Option`. For these reasons, I would say that we should keep this and remove `get_default`. 

As an examination of this in more depth:

If `T` were `Any` or `Resource` (e.g. if we had the following function declaration: `fun x<T>(t: &Option<T>, dflt: T, ....)`):
* If we kept `get_default` but removed `borrow_default` we would be forced within this function to unpack `t` as follows:
```rust
let x = if (Option::is_none(t)) {
   &dflt
} else {
   Option::borrow(t)
};
```
* If we kept `borrow_default` but removed `get_default`:
```rust
let x = Option::borrow_default(t, &dflt);
```

In the case where `T` is `copyable` (`fun x<T: copyable>(t: &Option<T>, dflt: T, ...)`) we have:
* If we got rid of `get_default` but keep `borrow_default`:
```rust
let t = *Option::borrow_default(t, &dflt);
```
* If we got rid of `borrow_default` but kept `get_default`:
```rust
let t = Option::get_default(t, dflt);
```

Reply:
I agree with all of this except for the conclusion that we should eliminate `get_default` For the `let t = *Option::borrow_default(t, &dflt);` alternative, consider the (common) case where `dflt` is a literal. The source language will still let you write something like `&8`, but under the hood it's introducing a new local variable, a `StLoc` instruction to populate it, a `BorrowLoc` to pass it to the caller, then a `ReadRef` to match the `*`. That feels like a lot of extra work that the caller might prefer not to do (even if the syntax makes it easy...).